### PR TITLE
UCT/IB: Remove implicit on-demand paging

### PR DESCRIPTION
**Pull request by gleon99**
## What
Remove implicit ODP support. Explicit ODP is still on the place since it would be used for future ODPv2 support feature.

## Why ?
Implicit ODP support probably will have different API in the future.

### CODE REVIEW

**Comment by shamisp** [comment]
Can you please clarify why you suggest to remove it ? FYI, if you want to use RDMA with DAX in linux, ODP is only the functional option.

**Comment by shamisp** [comment]
@gleon99 - if the feature is functional, I don't see a good reason to rip it off. DAX is a part of upstream kernel and AFAIK the only functional way to interact between PMEM/DAX subsystem and RDMA subsystem of ODP. I'm sure there are other useful scenarios for ODP.

**Comment by gleon99** [comment]
@shamisp  implicit ODP is anyway not supported by DEVX, and we want to cleanup some legacy code.

**Comment by shamisp** [comment]
I had off-line chat with @yosefe and I'm okay with the ODP plan

File: src/uct/ib/mlx5/dv/ib_mlx5_ifc.h [hunk 1]
```diff
@@ -92,7 +92,6 @@ enum {
 
 enum {
     UCT_IB_MLX5_CAP_GENERAL   = 0x0,
-    UCT_IB_MLX5_CAP_ODP       = 0x2,
```
Comment (gleon99) [inline src/uct/ib/mlx5/dv/ib_mlx5_ifc.h:95]: Maybe keep it, for future ODP?

File: src/uct/ib/mlx5/dv/ib_mlx5_ifc.h [hunk 2]
```diff
@@ -509,40 +508,8 @@ struct uct_ib_mlx5_cmd_hca_cap_2_bits {
     uint8_t    reserved_at_220[0x6A0];
 };
 
-struct uct_ib_mlx5_odp_per_transport_service_cap_bits {
```
Comment (gleon99) [inline src/uct/ib/mlx5/dv/ib_mlx5_ifc.h:512]: Maybe keep?

File: src/uct/ib/mlx5/ib_mlx5.h [hunk 3]
```diff
@@ -54,7 +54,6 @@
 #define UCT_IB_MLX5_EXTENDED_UD_AV       0x80 /* htonl(0x80000000) */
 #define UCT_IB_MLX5_AV_GRH_PRESENT       0x40 /* htonl(UCS_BIT(30)) */
 #define UCT_IB_MLX5_BF_REG_SIZE          256
-#define UCT_IB_MLX5_CQE_VENDOR_SYND_ODP  0x93
```
Comment (gleon99) [inline src/uct/ib/mlx5/ib_mlx5.h:57]: Keep?

File: docs/source/glossaries.rst [hunk 4]
```diff
@@ -50,7 +50,6 @@ NAK         Negative Acknowledge
 NB          Non blocking operations
 NBI         Non blocking implicit
 NBX         Non blocking extended
-ODP         OnDemand Paging
```
Comment (yosefe) [inline docs/source/glossaries.rst:53]: no need to remove it

File: src/uct/ib/base/ib_md.c [hunk 5]
```diff
@@ -633,111 +613,13 @@ static uint64_t uct_ib_md_access_flags(uct_ib_md_t *md, unsigned flags,
...
-
     if (md->relaxed_order) {
         access_flags |= IBV_ACCESS_RELAXED_ORDERING;
     }
 
     return access_flags;
 }
 
-#if HAVE_NUMA
-static ucs_status_t uct_ib_mem_set_numa_policy(uct_ib_md_t *md, void *address,
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c:649]: can we remove numa includes?

File: src/uct/ib/mlx5/dv/ib_mlx5_ifc.h (hunk 1 above)
Comment (yosefe) [reply src/uct/ib/mlx5/dv/ib_mlx5_ifc.h:95]: let's remove; ODP would be driver API, not FW

File: src/uct/ib/mlx5/dv/ib_mlx5_ifc.h (hunk 2 above)
Comment (yosefe) [reply src/uct/ib/mlx5/dv/ib_mlx5_ifc.h:512]: remove - odp would be driver-level api

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 6]
```diff
@@ -89,8 +89,7 @@ uct_rc_mlx5_iface_check_rx_completion(uct_ib_iface_t   *ib_iface,
 
     if (((ecqe->op_own >> 4) == MLX5_CQE_RESP_ERR) &&
         (ecqe->syndrome == MLX5_CQE_SYNDROME_REMOTE_ABORTED_ERR) &&
-        ((ecqe->vendor_err_synd == UCT_IB_MLX5_CQE_VENDOR_SYND_ODP) ||
-         (ecqe->vendor_err_synd == UCT_IB_MLX5_CQE_VENDOR_SYND_PSN)))
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c:93]: keep

File: src/uct/ib/mlx5/ib_mlx5.h (hunk 3 above)
Comment (yosefe) [reply src/uct/ib/mlx5/ib_mlx5.h:57]: keep

**Comment by gleon99** [comment]
@yosefe fixed

**Review by yosefe — DISMISSED** [review]

File: src/uct/ib/base/ib_device.h [hunk 7]
```diff
@@ -90,8 +90,7 @@ enum {
     UCT_IB_DEVICE_FLAG_AV       = UCS_BIT(8),   /* Device supports compact AV */
     UCT_IB_DEVICE_FLAG_DC       = UCT_IB_DEVICE_FLAG_DC_V1 |
                                   UCT_IB_DEVICE_FLAG_DC_V2, /* Device supports DC */
-    UCT_IB_DEVICE_FLAG_ODP_IMPLICIT = UCS_BIT(9),
-    UCT_IB_DEVICE_FAILED        = UCS_BIT(10)   /* Got fatal error */
+    UCT_IB_DEVICE_FAILED        = UCS_BIT(9)   /* Got fatal error */
```
Comment (ivankochin) [inline src/uct/ib/base/ib_device.h:93]:
```suggestion
    UCT_IB_DEVICE_FAILED        = UCS_BIT(9)    /* Got fatal error */
```

**Comment by ivankochin** [comment]
@yosefe @brminich pls review

**Comment by gleon99** [comment]
@ivankochin maybe rephrase the PR description, explaining that we do retain some ODP-related code for future explicit ODP support.

File: test/gtest/uct/test_md.cc [hunk 8]
```diff
@@ -778,6 +778,7 @@ UCS_TEST_SKIP_COND_P(test_md, dereg_bad_arg,
         ASSERT_UCS_STATUS_EQ(UCS_ERR_UNSUPPORTED, status);
 
         params.field_mask = UCT_MD_MEM_DEREG_FIELD_MEMH;
+        /* coverity[double_free] */
```
Comment (brminich) [inline test/gtest/uct/test_md.cc:781]: why needed in this PR?

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 9]
```diff
@@ -3076,7 +3076,7 @@ class test_ucp_sockaddr_protocols_err_sender
     test_ucp_sockaddr_protocols_err_sender() {
         configure_peer_failure_settings();
         m_env.push_back(new ucs::scoped_setenv("UCX_IB_REG_METHODS",
-                                               "rcache,odp,direct"));
+                                               "rcache,direct"));
     }
```
Comment (brminich) [inline test/gtest/ucp/test_ucp_sockaddr.cc:3080]: do we really need this setting, given that `rcache,direct` is a default value?

File: test/gtest/uct/test_md.cc (hunk 8 above)
Comment (gleon99) [reply test/gtest/uct/test_md.cc:781]: Coverity gives false positives without it, I discovered when initially wrote this PR.
Without these suppressions, it thinks that the invalid `uct_md_mem_dereg_v2` calls really free the `memh`, thus it is "freed multiple times".

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 10]
```diff
@@ -90,7 +90,7 @@ uct_rc_mlx5_iface_check_rx_completion(uct_ib_iface_t   *ib_iface,
     if (((ecqe->op_own >> 4) == MLX5_CQE_RESP_ERR) &&
         (ecqe->syndrome == MLX5_CQE_SYNDROME_REMOTE_ABORTED_ERR) &&
         ((ecqe->vendor_err_synd == UCT_IB_MLX5_CQE_VENDOR_SYND_ODP) ||
-         (ecqe->vendor_err_synd == UCT_IB_MLX5_CQE_VENDOR_SYND_PSN)))
+        (ecqe->vendor_err_synd == UCT_IB_MLX5_CQE_VENDOR_SYND_PSN)))
```
Comment (ivankochin) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c:93]:
```suggestion
         (ecqe->vendor_err_synd == UCT_IB_MLX5_CQE_VENDOR_SYND_PSN)))
```

**Review by brminich — DISMISSED** [review]

**Comment by ivankochin** [comment]
@yosefe pls review

File: src/uct/ib/configure.m4 [hunk 11]
```diff
@@ -229,10 +229,7 @@ AS_IF([test "x$with_ib" = "xyes"],
            AC_DEFINE([HAVE_ODP], 1, [ODP support])
 
            AC_CHECK_DECLS(IBV_ODP_SUPPORT_IMPLICIT, [with_odp_i=yes], [],
-                          [[#include <infiniband/verbs.h>]])
-
-           AS_IF([test "x$with_odp_i" = "xyes" ], [
-               AC_DEFINE([HAVE_ODP_IMPLICIT], 1, [Implicit ODP support])])])
+                          [[#include <infiniband/verbs.h>]])])
```
Comment (yosefe) [inline src/uct/ib/configure.m4:232]: shall we remove lines 231-232 as well?

**Review by brminich — APPROVED** [review]

File: src/uct/ib/configure.m4 [hunk 12]
```diff
@@ -226,10 +226,7 @@ AS_IF([test "x$with_ib" = "xyes"],
                       [[#include <infiniband/verbs.h>]])
 
        AS_IF([test "x$with_odp" = "xyes" ], [
-           AC_DEFINE([HAVE_ODP], 1, [ODP support])
-
-           AC_CHECK_DECLS(IBV_ODP_SUPPORT_IMPLICIT, [with_odp_i=yes], [],
-                          [[#include <infiniband/verbs.h>]])])
+           AC_DEFINE([HAVE_ODP], 1, [ODP support])])
```
Comment (yosefe) [inline src/uct/ib/configure.m4:229]: I would remove "with_odp" option altogether. seems it's not useful for anything anymore. WDYT?

**Review by yosefe — DISMISSED** [review]

**Comment by ivankochin** [comment]
I got weird issue during the squashing, let's continue discussion about this patch here #8924

**Comment by yosefe** [comment]
https://github.com/openucx/ucx/compare/4d43fa..de71ac6

**Review by brminich — APPROVED** [review]

**Comment by ivankochin** [comment]
@yosefe pls approve
